Feature/detect usb connection

zach1590 opened the pull request.

zach1590 committed:
Improve logs and error handling

zach1590 committed:
detect when a usb is connected

Added ability to detect when a usb is connected. If the connected
device was an arduino automatically select it to open a port

zach1590 committed:
Remove line that would continue to monitor for connections

jeffstjean commented on usb/usb_tool.js:
```diff
@@ -44,12 +46,18 @@ const on_ws_message = (message) => {
             case 'COMMAND':
                 handle_command(message.command, message.arguments);
                 break;
+
+            // invalid message.type (let the backend know it messed up)
+            default:
+                ws.send(JSON.stringify({error: "The command contains an invalid type"}));
+
```
Cheers for this thank you

jeffstjean commented on usb/usb_tool.js:
```diff
@@ -59,6 +67,24 @@ const ws_connect = () => {
 ws_connect()
 
+usbDetect.on('add', function(usb_port) { 
+    // The usb_port object does not have the path field needed to open
+    // a connection to the port so instead check if the manufacturer
+    // is correct, and if it is find which port its on and select it.
+    // This could go wrong if 2 arduinos are plugged in
```
Can we provide some checking to see if two potential devices are connected? I did this [previously](https://github.com/UWO-Aero-Design/gnd-station/blob/03b9f27fff5ae569a1404c29db7bba48708652b3/api/config/usb.js#L72-L85) by creating a list of candidate devices and only attempting an auto connect if the list had exactly one entry.

zach1590 replied:
I change that to

jeffstjean commented on usb/usb_tool.js:
```diff
@@ -18,10 +19,11 @@ const ws = new WebSocket(`ws://localhost:${port}`);
 
 let device = null;
 let parser = null;
+let is_arduino = false;
```
This doesn't appear to be used anywhere. is it no longer used after my message regarding not worrying about Arduinos? If not used, it can be deleted.

If it's going to be used, can we either check to see if the current device _is_ an Arduino just prior to doing some Arduino-specific code or find another solution? Just trying to avoid too many globals.

zach1590 replied:
yeah, Ill get rid of it. Originally I had a different idea, and forgot about it

jeffstjean replied:
Sounds good. Just push the changes and re-request review and it should be good to go. Afterwards, can you do a quick review of #47? I can walk through it all if you're interested but more so just checking if there's anything major / any suggestions? I already have one for myself: more multi-line commit messages.

zach1590 committed:
reduce duplication

jeffstjean committed:
fixed usb-detection import

jeffstjean committed:
Added small delay to startup routine

Appears that usb-detection detects new devices before SerialPort's internal list updates
Not a pretty solution but adding a 500ms delay before the auto-connect routine resolved this on MacOS
Will have to determine if this is a bigger issue and if there's a better solution

jeffstjean commented:
Just had to make two small changes after testing it with actual hardware
